specs/networking/p2p-interface.md

```diff
@@ -0,0 +1,760 @@
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+# Network Fundamentals
+
+This section outlines the specification for the networking stack in Ethereum 2.0 clients.
+
+Sections that have differing parameters for mainnet launch and interoperability testing are split into subsections. Sections that are not split have the same parameters for interoperability testing as mainnet launch.
+
+## Transport
+
+Even though libp2p is a multi-transport stack (designed to listen on multiple simultaneous transports and endpoints transparently), we hereby define a profile for basic interoperability.
+
+#### Interop
+
+All implementations MUST support the TCP libp2p transport, and it MUST be enabled for both dialing and listening (i.e. outbound and inbound connections).
+
+The libp2p TCP transport supports listening on IPv4 and IPv6 addresses (and on multiple simultaneously). Clients SHOULD allow the operator to configure the listen IP addresses and ports, including the addressing schemes (IPv4, IPv6).
+
+To facilitate connectivity, and avert possible IPv6 routability/support issues, clients participating in the interoperability testnet MUST expose at least ONE IPv4 endpoint.
+
+All listening endpoints must be publicly dialable, and thus not rely on libp2p circuit relay, AutoNAT or AutoRelay facilities.
+
+Nodes operating behind a NAT, or otherwise undialable by default (e.g. container runtime, firewall, etc.), MUST have their infrastructure configured to enable inbound traffic on the announced public listening endpoint.
+
+#### Mainnet
+
+All requirements from the interoperability testnet apply, except for the IPv4 addressing scheme requirement.
+
+At this stage, clients are licensed to drop IPv4 support if they wish to do so, cognizant of the potential disadvantages in terms of Internet-wide routability/support. Clients MAY choose to listen only on IPv6, but MUST retain capability to dial both IPv4 and IPv6 addresses.
+
+Usage of circuit relay, AutoNAT or AutoRelay will be specifically re-examined closer to the time.
+
+## Encryption and identification
+
+#### Interop
+
+[SecIO](https://github.com/libp2p/specs/tree/master/secio) with `secp256k1` identities will be used for initial interoperability testing.
+
+The following SecIO parameters MUST be supported by all stacks:
+
+-  Key agreement: ECDH-P256.
+-  Cipher: AES-128.
+-  Digest: SHA-256.
+
+#### Mainnet
+
+[Noise Framework](http://www.noiseprotocol.org/) handshakes will be used for mainnet. libp2p Noise support [is in the process of being standardised](https://github.com/libp2p/specs/issues/195) in the libp2p project.
+
+Noise support will presumably include IX, IK and XX handshake patterns, and may rely on Curve25519 keys, ChaCha20 and Poly1305 ciphers, and SHA-256 as a hash function. These aspects are being actively debated in the referenced issue [ETH 2.0 implementers are welcome to comment and contribute to the discussion.]
```

Comment:
How would this interact with QUIC where the standard defines it's own authentication and confidentiality layer?

Reply:
Also it seems weird to me to support more than one noise handshake pattern.

Reply:
Also if you want to mandate an authenticated handshake like `XX`, how does this interact with `secp256k1` identity key?

Reply:
There is an extension to the noise protocol that supports signatures that may be relevant but to my knowledge none of the noise implementations support it.

https://github.com/noiseprotocol/noise_sig_spec/blob/master/output/noise_sig.pdf

Reply:
These are excellent points but belong in https://github.com/libp2p/specs/issues/195

Re QUIC, libp2p will understand the features that quic offers and disable its own corresponding protocols accordingly

Reply:
@zmanian this should cover most questions: https://github.com/libp2p/specs/issues/195#issuecomment-517353941

Offering ?K and ?X variants is useful because in some cases you will have the peer’s public static key (e.g. bootstrappers, prev connection). And, as the initiator, you’ll want to leverage it to encrypt your 0-RTT early data.

re: QUIC. Transports offering native encryption/authentication or multiplexing don’t overlay such features.